### ISSUE TITLE
New Cop: filter_map for Ruby > 2.7 instead of select + map

### ISSUE THREAD

**Issue by guillaumebriday** [rubocop-hq/rubocop#7946]
## Is your feature request related to a problem? Please describe.

That would be nice if Rubocop could detect the new `filter_map` method released with 2.7.

## Describe the solution you'd like

So instead of doing something like that:

```ruby
[1, 2, 3].select { |x| x.odd? }.map { |x| x.to_s }
```

Now we can write:

```ruby
[1, 2, 3].filter_map {|x| x.odd? ? x.to_s : nil }
```
## Additional context

More about the `filter_map` method here: [https://blog.saeloun.com/2019/05/25/ruby-2-7-enumerable-filter-map.html](https://blog.saeloun.com/2019/05/25/ruby-2-7-enumerable-filter-map.html)

**Comment by tejasbubane** [comment]
As discussed in https://github.com/rubocop-hq/ruby-style-guide/pull/819 this cop should be added to [rubocop-performance](https://github.com/rubocop-hq/rubocop-performance).

**Comment by koic** [comment]
Is `filter_map` obviously faster?

The following is an example:

```ruby
% cat filter_map.rb
require 'benchmark/ips'

ary = (1..1_000_000).to_a

Benchmark.ips do |x|
  x.report('select.map') { ary.select { |x| x.odd? }.map { |x| x.to_s } }
  x.report('filter_map') { ary.filter_map {|x| x.odd? ? x.to_s : nil } }
  x.compare!
end
```

Almost always the same result.

```console
% ruby filter_map.rb
Warming up --------------------------------------
          select.map     1.000  i/100ms
          filter_map     1.000  i/100ms
Calculating -------------------------------------
          select.map      4.755  (±21.0%) i/s -     24.000  in   5.163804s
          filter_map      5.029  (± 0.0%) i/s -     26.000  in   5.228925s

Comparison:
          filter_map:        5.0 i/s
          select.map:        4.8 i/s - same-ish: difference falls within error
```

I also saw the result of slow speed.

```console
% ruby filter_map.rb
Warming up --------------------------------------
          select.map     1.000  i/100ms
          filter_map     1.000  i/100ms
Calculating -------------------------------------
          select.map      5.317  (± 0.0%) i/s -     27.000  in   5.096556s
          filter_map      5.110  (± 0.0%) i/s -     26.000  in   5.138993s

Comparison:
          select.map:        5.3 i/s
          filter_map:        5.1 i/s - 1.04x  (± 0.00) slower
```

**Comment by tejasbubane** [comment]
@koic Yes, even I found skewed results for larger arrays. But for smaller ones, I found `filter_map` to be consistently faster:

```
                               user     system      total        real
100 times filter + map     0.000551   0.000045   0.000596 (  0.000653)
100 times filter_map       0.000458   0.000034   0.000492 (  0.000526)
10k times - filter + map  32.848246   0.302623  33.150869 ( 34.061939)
10k times - filter_map    32.497194   0.344086  32.841280 ( 34.530685)
```

**Comment by koic** [comment]
@tejasbubane Can you show me the benchmark code?

**Comment by tejasbubane** [comment]
```ruby
# frozen_string_literal: true

require 'benchmark'

numbers_10 = (1..10).to_a.shuffle
numbers_10k = (1..10_000).to_a.shuffle

Benchmark.bmbm do |x|
  x.report('100 times filter + map') do
    100.times do
      numbers_10.filter { |n| n.odd? }.map { |n| n.to_s }
    end
  end

  x.report('100 times filter_map') do
    100.times do
      numbers_10.filter_map { |n| n.to_s if n.odd? }
    end
  end

  x.report('10k times - filter + map') do
    10_000.times do
      numbers_10k.filter { |n| n.odd? }.map { |n| n.to_s }
    end
  end

  x.report('10k times - filter_map') do
    10_000.times do
      numbers_10k.filter_map { |n| n.to_s if n.odd? }
    end
  end
end
```

**Comment by marcandre** [comment]
Note that not all `filter{...}.map{...}` can be changed into `filter_map{...}`: if `map{...}` can potentially produce a `nil` then no `filter_map` can be equivalent.

Finally, I'd be very curious to see if `numbers.filter(&:odd?).map(&:to_s)` is not faster than any `filter_map` one can write.

**Comment by guillaumebriday** [comment]
Up

**Comment by koic** [comment]
I will transfer this feature to rubocop -performance due to this is a dup with https://github.com/rubocop-hq/rubocop-performance/issues/178.
